File: 2022/Day 3/day3.py
"""
<3 python for this one
"""
import sys
import time
import math
import pathlib
from collections import defaultdict, Counter, deque
from copy import deepcopy

PATH="/Users/fredrikburmester/Documents/GitHub/advent-of-code"
DAY=3
INPUT_PATH=f"{PATH}/2022/Day {DAY}/input.txt"

sys.path.append(PATH)  
from utils.helpers import *

letters = "abcdefghijklmnopqrstuvwxyz"
capital_letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

def part1(input_list):
    """ Part 1"""

    total = 0

    for line in input_list:
        line = [str(x) for x in line]

        first_half = line[:len(line)//2]
        second_half = line[len(line)//2:]

        matching = [x for x in first_half if x in second_half]

        if len(matching) > 0:
            l = matching[0]
            if l in letters:
                index = letters.index(l) + 1
                total += index
                continue
            if l in capital_letters:
                index_2 = capital_letters.index(l) + 27
                total += index_2
                continue

    return total


def part2(input_list):
    """ Part 2 """

    total = 0

    # every third line
    for i in range(0, len(input_list), 3):

        line = [str(x) for x in input_list[i]]
        line_2 = [str(x) for x in input_list[i+1]]
        line_3 = [str(x) for x in input_list[i+2]]

        matching = [x for x in line if x in line_2 and x in line_3]

        if len(matching) > 0:
            l = matching[0]
            if l in letters:
                index = letters.index(l) + 1
                total += index
                continue
            if l in capital_letters:
                index_2 = capital_letters.index(l) + 27
                total += index_2
                continue

    return total


def main():
    with open(INPUT_PATH, "r", encoding='UTF-8') as file:
        input_list = [str(line.strip()) for line in file]

    t0 = time.time()
    result = part1(input_list)
    t1 = time.time()
    print(f"\n{bcolors.OKGREEN}{result}{bcolors.ENDC} is the result of part 1.\n{bcolors.OKBLUE}{round(t1-t0, 4)}{bcolors.ENDC} seconds\n")

    t0 = time.time()
    result2 = part2(input_list)
    t1 = time.time()
    print(f"\n{bcolors.OKGREEN}{result2}{bcolors.ENDC} is the result of part 2.\n{bcolors.OKBLUE}{round(t1-t0, 4)}{bcolors.ENDC} seconds\n")


if __name__ == "__main__":
    main()

File: 2022/Day 3/day3.go
// Only part one for now.

package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func get_value(l string) int {
	letters := "abcdefghijklmnopqrstuvwxyz"
	capital_letters := "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

	for j := 0; j < len(letters); j++ {
		if l == string(letters[j]) {
			return j + 1
		}
	}
	for j := 0; j < len(capital_letters); j++ {
		if l == string(capital_letters[j]) {
			return j + 27
		}
	}
	return 0
}

func part1(input_file []string) int {
	total := 0

	for i := 0; i < len(input_file); i++ {
		line := input_file[i]
		length := len(line)

		first_half := line[0 : length/2]
		second_half := line[length/2 : length]

		match := ""

		for j := 0; j < len(first_half); j++ {
			for k := 0; k < len(second_half); k++ {
				if first_half[j] == second_half[k] {
					match = string(first_half[j])
					break
				}
			}
		}

		total += get_value(match)
	}

	return total
}

func part2(input_file []string) int {
	return 0
}

func main() {
	f, err := os.Open("./input.txt")
	check(err)

	fileScanner := bufio.NewScanner(f)
	fileScanner.Split(bufio.ScanLines)
	array_of_strings := []string{}

	for fileScanner.Scan() {
		array_of_strings = append(array_of_strings, fileScanner.Text())
	}
	start := time.Now()
	fmt.Println(part1(array_of_strings))
	elapsed := time.Since(start)

	fmt.Printf("Part 1 took %s \n", elapsed)

	start = time.Now()
	fmt.Println(part2(array_of_strings))
	elapsed = time.Since(start)

	fmt.Printf("Part 2 took %s \n", elapsed)

	f.Close()
}
